### ISSUE TITLE
Support for Litespeed Cache (LSCache)

### ISSUE THREAD

**Issue by eurobank** [osclass/Osclass#2182]
Most of scripts support this cache. It would be great if we can also have it for Osclass. Via a plugin or a simpler via .htaccess. 

All the information is here (with examples):

https://www.litespeedtech.com/support/wiki/doku.php/litespeed_wiki:cache:no-plugin-setup-guidline

**Comment by conejoninja** [comment]
Never used litespeed before, but if I'm reading this correctly, is the cache only available in paid versions (LiteSpeed Enterprise Edition) of the webserver?

**Comment by eurobank** [comment]
I have no idea. A lot of hosting companies use LiteSpeed webserver for many reasons. So it came with my shared plan as a bonus. 

Wordpress already has a plugin for it. 

https://wordpress.org/plugins/litespeed-cache/

I read on their wiki "LSCache: LiteSpeed Cache, our extremely powerful page caching feature, is bundled free with 2-CPU or greater licenses. For all other licenses, LSCache can be bought as a separate module."

Some information:

https://www.litespeedtech.com/support/wiki/doku.php/litespeed_wiki:cache:no-plugin-setup-guidline

https://www.litespeedtech.com/support/wiki/doku.php/litespeed_wiki:cache:no-plugin-wordpress

**Comment by eurobank** [comment]
Also i tried to just make a .htaccess code for it (following the example) but had many problems that i don't know how to solve.

Login/Register for example doesn't seem to work. Some cookies code is needed, exclusions etc etc.

**Comment by dev-101** [comment]
If it works like Varnish (from their docs), than Varnish doesn't work well with anything that has cookies, and Osclass depends on them.

**Comment by conejoninja** [comment]
Any caching on a logged user is complicated, as the page is unique to each user. "Non-logged" visits are easier, but if you are using cookies or session to store some sort of data (like default language, if you have accepted tracking cookies or not,...) then you might have a problem.


I'm more worried that litespeed looks like a paid server, it could come free with some hostings, but a) said hostings are probably paid too and b) I would highly appreciate if we could work/develop on a local environment instead of some external server

**Comment by eurobank** [comment]
@dev-101 Does Wordpress has cookies ? Because it works wonderfully with Wordpress. I ditched all other caching plugins. 

Does this informatin helps your question:

"Our cache plugin will not trigger PHP. It is very similar to html page caches, except the benefit of LiteSpeed Cache is the server is aware of how the cached pages are stored.

IThe server stores the cache pages as static files and keeps a hash table with extra information about each stored page. The extra information allows us to purge pages more effectively as well as allows us to further improve the cache if needed."

Also if anyone wants to do ANY kind of test, i can offer a full account with ftp on a subdomain.

**Comment by eurobank** [comment]
@daniel maybe you should contact them directly. They are very helpful and friendly. 

Maybe they could give you some license. I dunno.

**Comment by eurobank** [comment]
@dev101 

I see that plugins are available for Wordpress, Joomla, vBulletin, Mediawiki, IP Board, vbSeo. I find it hard to believe it will not work well with Osclass.

**Comment by dev-101** [comment]
True, there are. But, personally, I have very little interest in it.

**Comment by eurobank** [comment]
How can i see if a user or admin is logged so i can disable caching ? I'm told to check cookies but i don't see any new cookie set when logged in.

**Comment by conejoninja** [comment]
osc_is_web_user_logged_in() and osc_is_admin_user_logged_in() might help here

**Comment by eurobank** [comment]
Thanks. I wasn't clear in my question. I must do all the work in .htaccess via mod_rewrite. Something like:

```
#To exclude users after login you will need to see if cookies are being set if so add the following and replace with the cookie that is being set
RewriteCond %{HTTP_COOKIE} !logged|cart [NC] 
```

**Comment by conejoninja** [comment]
https://github.com/osclass/Osclass/blob/master/oc-includes/osclass/helpers/hUsers.php#L66
oc_userId and oc_userSecret cookies should exists and be non empty

But cookies are only there for the "remember me" option. A logged user information resides on the session, not sure how you could access that from the .htaccess. Might be useful to look at the vbulletin plugin or something like that.

**Comment by eurobank** [comment]
It seems Vbulletin has cookies:

```
## select which pages to cache
RewriteCond %{HTTP_COOKIE} !bbimloggedin=yes
RewriteCond %{HTTP_COOKIE} !bbuserid=
RewriteCond %{HTTP_COOKIE} !bbpassword=
```

**Comment by eurobank** [comment]
I see how this works in this old cache plugin:

https://github.com/PluginCache-osclass/plugincache

I will try to find someone to "adapt" this plugin to work with Litespeed cache. I shouldn't be hard to do.

**Comment by Tangol** [comment]
@conejoninja Please note that LiteSpeed also has a free version: http://open.litespeedtech.com/mediawiki/

**Comment by eurobank** [comment]
@Tangol Yes there is a free version of the web server but the LSCaching is bundled with the paid only. I think.

**Comment by Tangol** [comment]
@eurobank http://open.litespeedtech.com/mediawiki/index.php/Help:Features
**Main Features**
- Rewrite engine uses Apache mod_rewrite syntax.
- High-performance **page caching**.
- Support of third-party modules through API (LSIAPI).

**Comment by eurobank** [comment]
@Tangol I don't think so. OpenLiteSpeed has a similar product.

http://openlitespeed.com/threads/does-ols-includes-lscache.1882/

"OLS does not include LSCache. However, it does include the cache module, which functions very similarly."
